### PR TITLE
Use MBED_APP_* and MBED_RAM_* in linker script

### PR DESCRIPTION
**Pull request by mikisch81**
## Linker work

- Secure/Non-secure/Shared addresses are configuration params.
- They are sent to the compiler and used as constant MACROs.
- They are sent to the linker as MBED_APP_START/SIZE and MBED_RAM_START/SIZE.

---
This change is [Reviewable](https://reviewable.io/reviews/kfnta/mbed-os/70)

### CODE REVIEW

File: tools/toolchains/__init__.py [hunk 1]
```diff
@@ -757,6 +757,29 @@ def add_linker_defines(self):
             self.ld.append(define_string)
             self.flags["ld"].append(define_string)
 
+        if Target.is_secure_target(self.target.name) or Target.is_non_secure_target(self.target.name):
```
Comment (orenc17) [inline tools/toolchains/__init__.py]:
```suggestion
        flags2params = {}
        if Target.is_non_secure_target(self.target.name):
            flags2params = {
                "MBED_APP_START": "target.non-secure-flash-start",
                "MBED_APP_SIZE": "target.non-secure-flash-size",
                "MBED_RAM_START": "target.non-secure-ram-start",
                "MBED_RAM_SIZE": "target.non-secure-ram-size"
            }
        
        if Target.is_secure_target(self.target.name):
            flags2params = {
                "MBED_APP_START": "target.secure-flash-start",
                "MBED_APP_SIZE": "target.secure-flash-size",
                "MBED_RAM_START": "target.secure-ram-start",
                "MBED_RAM_SIZE": "target.secure-ram-size"
            }
        
        for flag, param in flags2params.items():
            define_string = self.make_ld_define(flag, params[param].value)
            self.ld.append(define_string)
            self.flags["ld"].append(define_string)
```

File: tools/targets/__init__.py [hunk 2]
```diff
@@ -336,6 +336,10 @@ def labels(self):
     def is_secure_target(mcu):
```
Comment (orenc17) [inline tools/targets/__init__.py]:
```suggestion
    def is_PSA_secure_target(mcu):
```

File: tools/targets/__init__.py [hunk 3]
```diff
@@ -336,6 +336,10 @@ def labels(self):
...
         return 'SPE_Target' in Target.get_target(mcu).labels
 
+    @staticmethod
+    def is_non_secure_target(mcu):
```
Comment (orenc17) [inline tools/targets/__init__.py]:
```suggestion
    def is_PSA_non_secure_target(mcu):
```

File: components/psa/spm/COMPONENT_SPE/spm_init.c [hunk 4]
```diff
@@ -97,15 +97,15 @@ void psa_spm_init(void)
...
-    size_t   non_secure_flash_len  = spm_hal_get_plat_non_sec_flash_len();
+    uint32_t secure_ram_base   = PSA_SECURE_RAM_START;
+    size_t   secure_ram_len    = PSA_SECURE_RAM_SIZE;
+    uint32_t secure_flash_base = PSA_SECURE_FLASH_START;
+    size_t   secure_flash_len  = PSA_SECURE_FLASH_SIZE;
+
+    uint32_t non_secure_ram_base   = PSA_NON_SECURE_RAM_START;
+    size_t   non_secure_ram_len    = PSA_NON_SECURE_RAM_SIZE;
+    uint32_t non_secure_flash_base = PSA_NON_SECURE_FLASH_START;
+    size_t   non_secure_flash_len  = PSA_NON_SECURE_FLASH_SIZE;
```
Comment (alzix) [inline components/psa/spm/COMPONENT_SPE/spm_init.c]: i suggest changing all the checks that follow this line to a static asserts

Comment (mikisch81) [reply components/psa/spm/COMPONENT_SPE/spm_init.c]: I know, can you guys help me with that?

Comment (alzix) [reply components/psa/spm/COMPONENT_SPE/spm_init.c]: with pleasure!!!! let me handle that :)

**Comment by mikisch81** [comment]
@orenc17 @alzix 
Compilation is successfull using ARM compiler but it does not run, need to investigate..

**Comment by alzix** [comment]
@mikisch81 , please note that your code does not compile :(
```
Traceback (most recent call last):
  File "mbed-os\tools\make.py", line 310, in <module>
    if Target.is_secure_target(mcu):
AttributeError: type object 'Target' has no attribute 'is_secure_target'
```

File: components/psa/spm/COMPONENT_SPE/spm_init.c [hunk 5]
```diff
@@ -27,6 +27,66 @@ MBED_STATIC_ASSERT( MBED_CONF_SPM_IPC_MAX_NUM_OF_MESSAGES <= PSA_HANDLE_MGR_MAX_
...
+
+MBED_STATIC_ASSERT(
+    (PSA_NON_SECURE_RAM_START + PSA_NON_SECURE_RAM_SIZE - 1) > PSA_NON_SECURE_RAM_START,
+    "Illegal non-secure ram region!\n"
+);
+
+MBED_STATIC_ASSERT(
+    (PSA_NON_SECURE_FLASH_START + PSA_NON_SECURE_FLASH_SIZE - 1) > PSA_NON_SECURE_FLASH_START,
+    "Illegal non-secure flash region!\n"
+);
```
Comment (danny4478) [inline components/psa/spm/COMPONENT_SPE/spm_init.c]:
```suggestion
);

MBED_STATIC_ASSERT(
    (PSA_SHARED_RAM_START + PSA_SHARED_RAM_SIZE - 1) > PSA_SHARED_RAM_START,
    "Illegal shared ram region!\n"
);

```

File: components/psa/spm/COMPONENT_SPE/spm_init.c [hunk 6]
```diff
@@ -27,6 +27,66 @@ MBED_STATIC_ASSERT( MBED_CONF_SPM_IPC_MAX_NUM_OF_MESSAGES <= PSA_HANDLE_MGR_MAX_
...
+    )
+
+OVERLAP_CHECK(
+    "PSA_SECURE_RAM",     PSA_SECURE_RAM_START,     PSA_SECURE_RAM_SIZE,
+    "PSA_NON_SECURE_RAM", PSA_NON_SECURE_RAM_START, PSA_NON_SECURE_RAM_SIZE,
+);
+
+OVERLAP_CHECK(
+    "PSA_SECURE_RAM", PSA_SECURE_RAM_START, PSA_SECURE_RAM_SIZE,
+    "PSA_SHARED_RAM", PSA_SHARED_RAM_START, PSA_SHARED_RAM_START,
```
Comment (danny4478) [inline components/psa/spm/COMPONENT_SPE/spm_init.c]:
```suggestion
    "PSA_SHARED_RAM", PSA_SHARED_RAM_START, PSA_SHARED_RAM_SIZE,
```

File: components/psa/spm/COMPONENT_SPE/spm_init.c [hunk 7]
```diff
@@ -27,6 +27,66 @@ MBED_STATIC_ASSERT( MBED_CONF_SPM_IPC_MAX_NUM_OF_MESSAGES <= PSA_HANDLE_MGR_MAX_
...
+);
+
+OVERLAP_CHECK(
+    "PSA_SHARED_RAM",     PSA_SHARED_RAM_START,     PSA_SHARED_RAM_START,
```
Comment (danny4478) [inline components/psa/spm/COMPONENT_SPE/spm_init.c]:
```suggestion
    "PSA_SHARED_RAM",     PSA_SHARED_RAM_START,     PSA_SHARED_RAM_SIZE,
```

File: components/psa/spm/COMPONENT_SPE/spm_init.c [hunk 8]
```diff
@@ -27,6 +27,66 @@ MBED_STATIC_ASSERT( MBED_CONF_SPM_IPC_MAX_NUM_OF_MESSAGES <= PSA_HANDLE_MGR_MAX_
...
+    "PSA_NON_SECURE_RAM", PSA_NON_SECURE_RAM_START, PSA_NON_SECURE_RAM_SIZE,
+);
+
+OVERLAP_CHECK(
+    "PSA_SECURE_FLASH",     PSA_SECURE_FLASH_START, PSA_SECURE_FLASH_SIZE,
+    "PSA_NON_SECURE_FLASH", PSA_NON_SECURE_FLASH_START, PSA_NON_SECURE_FLASH_SIZE
+);
+
+
```
Comment (danny4478) [inline components/psa/spm/COMPONENT_SPE/spm_init.c]: I will go for full checks. Currently flash and ram overlap is not checked.
If we don't check it on purpose, we should add a comment with the reason for it.
Something like:
```suggestion
// Flash and RAM are assumed to be far from each other - overlap checks are not done
```

File: components/psa/spm/COMPONENT_SPM_MAILBOX/COMPONENT_SPE/spm_mailbox_spe.c [hunk 9]
```diff
@@ -103,22 +103,20 @@ void on_popped_item(ipc_queue_item_t item)
...
     // This struct is set with initial values for the address table (addresses of CM0+ / CM4 shared memory)
     const addr_table_t shared_addr_table = {
         .magic        = ADDR_TABLE_MAGIC,
-        .tx_queue_ptr = (uintptr_t)( &__shared_memory_start +
-                                     (sizeof(addr_table_t) / sizeof(uint32_t))
-                                   ),
-        .rx_queue_ptr = (uintptr_t)( &__shared_memory_start +
-                                     ((sizeof(addr_table_t) + sizeof(ipc_base_queue_t)) / sizeof(uint32_t))
-                                   )
+        .tx_queue_ptr = (uintptr_t)( PSA_SHARED_RAM_START + sizeof(addr_table_t) ),
```
Comment (danny4478) [inline components/psa/spm/COMPONENT_SPM_MAILBOX/COMPONENT_SPE/spm_mailbox_spe.c]: The original dividing by sizeof(uint32_t) was verified using a debugger.
If this is modified now, it is better to re-check with the debugger to make sure we do not have a miscalculation here - this is a critical place.

**Pull request merged**